### ISSUE TITLE
Segfault when closing curl handle from callback

### ISSUE THREAD

**Issue by y3llowcake** [facebook/hhvm#8131]
### HHVM Version
```
cy@slack-dev147:~$ hhvm --version
HipHop VM 3.21.6 (rel)
Compiler: remotes/origin/3.21.6-slack-0-g5456e6ccc5335f64a2c69c51af4502ef7ccae030
Repo schema: 7cba64a50624349c19446a30c2d9600881553820
```
 - 3.21.6
 - 3.24.2
 - nightly: 20180221

### Operating System and Version
```
cy@slack-dev147:~$ uname -a
Linux slack-dev147 4.4.0-1049-aws #58-Ubuntu SMP Fri Jan 12 23:17:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
cy@slack-dev147:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial
```

+ hhvm/hhvm docker images

### Standalone code, or other way to reproduce the problem
```
cy@slack-dev147:~$ cat derp.php
<?php
$handle = curl_init("https://www.facebook.com");
curl_setopt($handle, CURLOPT_HEADERFUNCTION, function($handle, $header) {
        curl_close($handle);
});
curl_exec($handle);
```
### Expected result
Not segfault.
### Actual result
```
cy@slack-dev147:~$ hhvm derp.php                                                                                        
Core dumped: Segmentation fault                                                                                         
Stack trace in /tmp/stacktrace.27325.log                                                                                
Segmentation fault

cy@slack-dev147:~$ php5 derp.php                                                                                                                                                                                                                                                                                                                                                                                                                                                                

Warning: curl_close(): Attempt to close cURL handle from a callback in /home/cy/derp.php on line 4                      

Call Stack:                                                                                                             
    0.0001     229544   1. {main}() /home/cy/derp.php:0                                                                 
    0.0006     231240   2. curl_exec() /home/cy/derp.php:6                                                              
    0.0823     231408   3. {closure:/home/cy/derp.php:3-5}() /home/cy/derp.php:6                                        
    0.0823     231456   4. curl_close() /home/cy/derp.php:4
```

**Comment by fredemmott** [comment]
Reproduced on everything through nightlies:

```
$ docker run -v $(pwd):/tmp/test --rm hhvm/hhvm:nightly hhvm /tmp/test/test.php
Core dumped: Segmentation fault
Stack trace in /tmp/stacktrace.1.log
```

**Comment by fredemmott** [comment]
Filed internal task: T26266727

**Comment by fredemmott** [comment]
@y3llowcake : thanks for the excellent report and repo case :)

**Comment by fredemmott** [comment]
https://github.com/facebook/hhvm/commit/90df62eeb0afd46ba6ecb2c9512c9babe092b998

**Comment by fredemmott** [comment]
Fixed in 3.21.7 and 3.24.3
